src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+        }
+        Ok(config)
+    }
+}
+
+/// An async-drop style `Guard` for `mongodb::Client`. Use this guard to hold a client,
+/// the `client::shutdown` is called in an async manner when the guard is dropped.
+/// Please be aware this is a "best effort" style shutdown, which may not be successful if the
+/// tokio runtime is in the process of terminating. However, the server-side resources will be
+/// cleaned up eventually due to the session expiration.
+/// see [this issue](https://github.com/mongodb/mongo-rust-driver/issues/719) for more information
+struct ClientGuard {
+    tx: tokio::sync::oneshot::Sender<()>,
+    client: Client,
+}
+
+impl ClientGuard {
+    fn new(name: String, client: Client) -> Self {
+        let client_copy = client.clone();
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        tokio::spawn(async move {
+            tracing::debug!(%name, "waiting for client to shut down");
+            let _ = rx.await;
+            tracing::debug!(%name, "sender dropped now calling client's shutdown");
+            // shutdown may stuck if the resources created by client are not dropped at this point.
+            // As recommended by [shutdown](https://docs.rs/mongodb/2.8.2/mongodb/struct.Client.html#method.shutdown)
+            // documentation, we should make our resources usage shorter-lived than the client. So if this happens,
+            // there are some programming error in our code.
+            client_copy.shutdown().await;
+            tracing::debug!(%name, "client shutdown succeeded");
+        });
+        Self { tx, client }
+    }
+}
+
+impl Deref for ClientGuard {
+    type Target = Client;
+
+    fn deref(&self) -> &Self::Target {
+        &self.client
+    }
+}
+
+#[derive(Debug)]
+pub struct MongodbSink {
+    pub config: MongodbConfig,
+    param: SinkParam,
+    schema: Schema,
+    pk_indices: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl MongodbSink {
+    pub fn new(param: SinkParam) -> Result<Self> {
+        let config = MongodbConfig::from_hashmap(param.properties.clone())?;
+        let pk_indices = param.downstream_pk.clone();
+        let is_append_only = param.sink_type.is_append_only();
+        let schema = param.schema();
+        Ok(Self {
+            config,
+            param,
+            schema,
+            pk_indices,
+            is_append_only,
+        })
+    }
+}
+
+impl TryFrom<SinkParam> for MongodbSink {
+    type Error = SinkError;
+
+    fn try_from(param: SinkParam) -> std::result::Result<Self, Self::Error> {
+        MongodbSink::new(param)
+    }
+}
+
+impl Sink for MongodbSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type LogSinker = LogSinkerOf<MongodbSinkWriter>;
+
+    const SINK_NAME: &'static str = MONGODB_SINK;
+
+    fn is_sink_decouple(desc: &SinkDesc, user_specified: &SinkDecouple) -> Result<bool> {
+        match user_specified {
+            SinkDecouple::Default => Ok(desc.sink_type.is_append_only()),
```

Comment:
Usually we enable sink decouple by default when the sink support async truncate, which means data can be delivered before checkpoint barrier comes. 

However, it seems that the mongodb sink writer only ensure delivery on checkpoint barrier. In this case, when sink decouple is enabled, almost all data will be written our internal log store, which causes implicit overhead. 

Therefore, it's better to set it as false by default.